sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -4724,13 +4748,20 @@ pi_result cuda_piextUSMEnqueuePrefetch(pi_queue queue, const void *ptr,
                                        const pi_event *events_waitlist,
                                        pi_event *event) {
 
-// CUDA has an issue with cuMemPrefetchAsync returning cudaErrorInvalidDevice
-// for Windows machines
-// TODO: Remove when fix is found
-#ifdef _MSC_VER
-  cl::sycl::detail::pi::die(
-      "cuda_piextUSMEnqueuePrefetch does not currently work on Windows");
-#endif
+  // Certain cuda devices and Windows do not have support for some Unified
+  // Memory features. cuMemPrefetchAsync requires concurrent memory access
+  // for managed memory. Therfore, ignore prefetch hint if concurrent managed
+  // memory access is not available.
+  int isConcurrentManagedAccessAvailable = 0;
+  cuDeviceGetAttribute(&isConcurrentManagedAccessAvailable,
+                       CU_DEVICE_ATTRIBUTE_CONCURRENT_MANAGED_ACCESS,
+                       queue->get_context()->get_device()->get());
+  if (!isConcurrentManagedAccessAvailable) {
+    setErrorMessage("Prefetch hint ignored as device does not support "
```

Comment:
Could you please clarify why can't we return a PI error code which would be interpreted on the SYCL RT side?
How can we guarantee that getLastError approach will correctly work in multithreaded environment?

Reply:
Sure, I think the main reason it does not make sense to return an error code is that it would be very specific to this backend. This error code would have to express that the operation failed due to the os/device not supporting concurrent memory and that this error is not a fail case but more of a performance warning as there is a redundant prefetch call. If the strategy of adding new error codes for these circumstances is taken for all backends, I would expect many error codes that would only be relevant to very specific circumstances and backends.

I think that this can be made thread-safe with mutex locks. However, I do think there may be circumstances where the wrong message is read for a given error.

Reply:
I support for adding this new PI API as it makes error-reporting easier () and such will make errors more clear to the end-user in the long run. A somewhat related example is that we are asked to to distinguish between plugin's internal errors vs. underlying native runtime errors. This infrastructure will save us from adding many new PI error codes.

For thread-safety we could block setting a new error until previous message is read (and the last error is cleared and mutex unlocked). This requires that PI_PLUGIN_SPECIFIC_ERROR would have to always be followed by a piPluginGetLastError to avoid dead-locks, but it is easy to document and implement with a wrapper function.

Reply:
Another approach to address thread safety issue is to use `thread_local` for messages storage, so each thread can read its own error message after PI call is done.

Reply:
I agree `thread_local` would be a tidy way of handling message storage with multiple threads. I have updated the PR to use it.